network/controller/bootstrap/session_manager.go

```diff
@@ -132,3 +172,74 @@ func (sm *SessionManager) ReleaseSession(id SessionID) (*Session, error) {
 	delete(sm.sessions, id)
 	return session, nil
 }
+
+func (sm *SessionManager) cleanupExpiredSessions() {
+	var sessionsByExpirationTime []*sessionWithID
+	for {
+		if len(sm.sessions) == 0 {
+			// Have no active sessions. Block till sessions will be added.
+			<-sm.newSessionNotification
+
+			sessionsByExpirationTime = sm.sortSessionsByExpirationTime()
+		}
+
+		// Get expiration time for next session and wait for it
+		nextSessionToExpire := sessionsByExpirationTime[0]
+		waitTime := nextSessionToExpire.expirationTime().Sub(time.Now())
+
+		select {
+		case <-sm.newSessionNotification:
+			// Handle new session. reorder expiration short list
+			sessionsByExpirationTime = sm.sortSessionsByExpirationTime()
+
+		case <-time.After(waitTime):
+			// Move forward through sessions and check whether we should delete the session
+			sessionsByExpirationTime = sm.expireSessions(sessionsByExpirationTime)
+
+		case <-sm.stopCleanupNotification:
+			return
+		}
+	}
+}
+
+func (sm *SessionManager) sortSessionsByExpirationTime() []*sessionWithID {
```

Comment:
do we need sorting? sessions have fixed duration, so they should be already ordered from less recent to most recent because we add them under lock one by one

Reply:
Yea, since TTL is stored inside session and latest session would need faster expiration than older one.